extensions/mssql/src/features.ts

```diff
@@ -921,3 +921,165 @@ export class SqlAssessmentServicesFeature extends SqlOpsFeature<undefined> {
 
 
 }
+
+export class ProfilerFeature extends SqlOpsFeature<undefined> {
+	private static readonly messagesTypes: RPCMessageType[] = [
+		contracts.StartProfilingRequest.type,
+		contracts.StopProfilingRequest.type,
+		contracts.ProfilerEventsAvailableNotification.type
+	];
+
+	constructor(client: SqlOpsDataClient) {
+		super(client, ProfilerFeature.messagesTypes);
+	}
+
+	public fillClientCapabilities(capabilities: ClientCapabilities): void {
+		//Utils.ensure(Utils.ensure(capabilities, 'connection')!, 'profiler')!.dynamicRegistration = true;
+	}
+
+	public initialize(capabilities: ServerCapabilities): void {
+		this.register(this.messages, {
+			id: UUID.generateUuid(),
+			registerOptions: undefined
+		});
+	}
+
+	protected registerProvider(options: undefined): Disposable {
+		const client = this._client;
+
+		let createSession = (ownerUri: string, sessionName: string, template: azdata.ProfilerSessionTemplate): Thenable<boolean> => {
+			let params: contracts.CreateXEventSessionParams = {
+				ownerUri,
+				sessionName,
+				template
+			};
+
+			return client.sendRequest(contracts.CreateXEventSessionRequest.type, params).then(
+				r => true,
+				e => {
+					client.logFailedRequest(contracts.CreateXEventSessionRequest.type, e);
+					return Promise.reject(e);
+				}
+			);
+		};
+
+		let startSession = (ownerUri: string, sessionName: string): Thenable<boolean> => {
```

Comment:
is there a fundamental different/preference for using "async method : Promise<>" vs "method: Thenable<>"?

Reply:
Thenable is only for the public extension APIs in azdata.d.ts - the reasoning be that VS Code wants to let extensions use their own promise libraries if they want.

Anywhere else you can (and should) use Promise.

Reply:
mssql.d.ts defines mostly with Thenable, with the exception being ISqlAssessmentService.  Is IsqlAssessmentService the "new right way" or "the old wrong way"?  Or does it depend on how we expect to expose this service to other extensions?
I sorta understand that Thenable means "i promise to act like a promise" vs "Promise is a specific node/typescript implementation of a promise". I don't want to rely on copy/pasting stuff I don't understand correctly.  How do I decide for ProfilerService which is correct?

Reply:
In mssql.d.ts I'd probably just stick with Thenable. Consistency in a public API is better (even though for the extension APIs I don't really see the point in that since these aren't really something I ever see outside customers using).